Rename package Chisel to chisel, add Chisel package for compatibility

ducky64 opened the pull request:
From a discussion a long time ago, `chisel` (lowercase c) is the preferred package name. Originally, since the rename was a breaking change, we were going to hold off, but it turns out creating a package object `Chisel` works for compatibility. So this shouldn't break any code, unless I forgot to include something in the compatibility object.

The compatibility object will not provide some of Chisel3's newer features, like access to IR types and other internals or library source locators (whenever we get around to those). Rename your imports to `chisel` to use those.

May break on machines with case insensitive file systems.

sdtwigg commented:
Instead of having users import chisel._, why not have them import something more curated (similar to how Chisel is now curated)?

sdtwigg commented on src/main/scala/chisel/compatibility.scala:
```diff
@@ -0,0 +1,139 @@
+
+  val BitPat = chisel.BitPat
+  type BitPat = chisel.BitPat
+
```
Element is missing.

ducky64 replied:
Fixed.

sdtwigg commented on src/main/scala/chisel/compatibility.scala:
```diff
@@ -0,0 +1,139 @@
+  type Pipe[T <: Data] = chisel.Pipe[T]
+}
+
+package Chisel {
```
You should be able to get away with having an object testers in the package object?

ducky64 replied:
Fixed now, I think.

sdtwigg commented on src/main/scala/chisel/compatibility.scala:
```diff
@@ -0,0 +1,139 @@
+// Allows legacy users to continue using Chisel (capital C) package name while
+// moving to the more standard package naming convention chisel (lowercase c).
+
+package object Chisel {
```
You need the implicit classes from the other package object in here too.

ducky64 replied:
Aren't those new chisel3 features? I think it would be nice to keep the new chisel3 features in `chisel`, otherwise we're going to end up maintaining this forever.

sdtwigg replied:
They were ported to chisel2.
On May 20, 2016 7:25 PM, "Richard Lin" <notifications@github.com> wrote:

> In src/main/scala/chisel/compatibility.scala
> <https://github.com/ucb-bar/chisel3/pull/197#discussion_r64126635>:
>
> > @@ -0,0 +1,139 @@
> > +// See LICENSE for license details.
> > +
> > +// Allows legacy users to continue using Chisel (capital C) package name while
> > +// moving to the more standard package naming convention chisel (lowercase c).
> > +
> > +package object Chisel {
>
> Aren't those new chisel3 features? I think it would be nice to keep the
> new chisel3 features in chisel, otherwise we're going to end up
> maintaining this forever.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/ucb-bar/chisel3/pull/197/files/d8381250d2f18d99edfd527a1353d5fa52fad10a#r64126635>
>

ducky64 replied:
Added.

ducky64 commented:
Good idea to curate chisel, but that's probably a larger change which merits more debate. What do you think should go in there?

sdtwigg commented:
My rationale for suggesting it now was that this change will already require people to update all of their imports. So, might as well have them update it to the 'curated' import.

ducky64 commented:
I guess that's fair. But still, what would you like to see in there? Maybe we also want to split chisel and chisel.utils?

sdtwigg commented:
Oh, it would probably have 'everything' in there (except the internal-ish stuff like: BackendCompilationUtilities, WhenContext, debug, and SwitchContext). This would also allow you to setup a 'more natural' package<->directory correspondence (which the chiselFrontend split broke).

Splitting out utils is also a good idea (but was unsure if you would be willing to do that change here as well). It is also unclear why certain things are even in 'core' (like BitPat, particularly)

sdtwigg commented:
Also, you can then consider moving the SourceInfo materialize into the chisel import (rather than the SourceInfo object) so that it is out of scope for the internals compile.

ducky64 commented:
This would probably be the only place to split out utils, since it would be a breaking change while allowing for the capital-C Chisel compatibility package.

BitPat is in core because it gets referenced in Bits (or at least something that's also in core). I don't think BitPat belongs in core, but otherwise the dependencies don't work. It might be possible to do something like implicit conversion from Bits to BitPatComparable (or something like that) that defines the relevant operations, though the experience with Chisel and implicits is generally kind of flaky.

sdtwigg commented:
It is only here:
https://github.com/ucb-bar/chisel3/blob/d742d70a05b5fa997517ea7b5eb2d15b23e7a431/chiselFrontend/src/main/scala/Chisel/Bits.scala#L467

and those methods can be added to Bits via extension methods.

ducky64 commented:
So turns out this will break chisel3 under Windows, presumably because of the case insensitive file  system. But this should be fine once we can drop the uppercase-C Chisel compatibility package.

Also, extension methods on Bits breaks rocket-chip (this is becoming a sadly common trend with Chisel and anything implicit). It seems like BitPat's implicit `apply` (conversion from UInt to BitPat) gets inserted (instead of the implicit conversion to the extension class), which promptly asserts out because it's not a literal.

Actually, having implicits that can assert out seems kind of sketchy in general...

sdtwigg commented:
The implicit conversions from UInt to/from BitPat is awful design.
An implicit conversion that can fail so easily should not be an implicit conversion. It just complicates error reporting and hurts abstraction guarantees.

The ultimate design of BitPat is so that users can do a partial equality check, right?
If further overloading of ===,=/= is so desired, I am highly tempted to investigate a variant of this solution: http://www.scala-lang.org/blog/2016/05/06/multiversal-equality.html
This would generally solve the problem and allow dropping the various implicit conversions to/from BitPat.

aswaterman commented:
Just to add context, the implicit conversions are there to ease porting
from Chisel2, where UInts can be bit-patterns and so the types are used
interchangeably.  The are, of course, a bad design.

A compromise might be to add them to a Chisel2 compatibility package, so
that their use is opt-in.

On Mon, May 23, 2016 at 3:21 PM, Stephen Twigg <notifications@github.com>
wrote:

> The implicit conversions from UInt to/from BitPat is awful design.
> An implicit conversion that can fail so easily should not be an implicit
> conversion. It just complicates error reporting and hurts abstraction
> guarantees.
>
> The ultimate design of BitPat is so that users can do a partial equality
> check, right?
> If further overloading of ===,=/= is so desired, I am highly tempted to
> investigate a variant of this solution:
> http://www.scala-lang.org/blog/2016/05/06/multiversal-equality.html
> This would generally solve the problem and allow dropping the various
> implicit conversions to/from BitPat.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ucb-bar/chisel3/pull/197#issuecomment-221113391>
>

ducky64 commented:
Removing the implicit conversion methods should make chisel3 safer and should allow extension methods on UInt to work with BitPat (so you can compare UInt === BitPat or BitPat === UInt without hardcoding knowledge of BitPat inside UInt). However, the compatibility layer would still not allow the proposed refactoring (remove BitPat functions from UInt and add them in through extension methods), since things would still break when the compatibility package is imported.

It might also be worth debating whether the BitPat to UInt function or the UInt to BitPat functions should exist at all, since they only work on a subset of possible cases and the places where they fail aren't captured by the type system. Is there a better way, style-wise, to express the equivalent use cases? Or perhaps Chisel's type system should be made more descriptive (like adding UIntLiteral)?

aswaterman commented:
Yeah, it's hard to imagine how to improve this without exposing the concept of Literal.

I don't think that it's inherently problematic that the conversions only work in some cases, though, since it's fairly intuitive when they're legal.

ducky64 commented:
I don't think it's always obvious where they're legal, you really only know where they're legal at where the UInt or BitPat is instantiated. If you have a function that takes in a general UInt or BitPat, it's no longer as clear and you have to leave it to a runtime check and function comments describing what inputs are acceptable and what aren't.

aswaterman commented:
I don't think that is inherently problematic, either.

ducky64 commented:
It's then hard to guarantee (at compile time) correctness of Chisel programs, since you have a leaky abstraction (UInt may be either a literal or not, but you can try to convert any UInt to a BitPat, and only at elaboration time will the program barf). It makes it more difficult to guarantee that a function taking a UInt and doing a cast is correct, you can only document it in the comments ("must take a literal!") and rely on elaboration-time asserts, the Scala compiler won't be able to help you.

aswaterman commented:
It's fundamentally impossible to guarantee at compile time the correctness of Chisel programs because of things like the interaction of width inference and out-of-bounds extraction.  IMO, this doesn't make the problem meaningfully worse.

I agree it would be ideal if there were UIntLiteral types, so this could be done more safely.  But users will still find ways to screw themselves -- for example, by dynamically casting UInt to UIntLiteral.

sdtwigg commented:
I think it is fine to have an explicit conversion that fails but not a implicit conversion that fails. The explicit conversion is just a function call with a domain that is more restricted than what you could encode in the type system.

ducky64 commented:
Still need a way to resolve this. We can keep the explicit conversions, but removing the implicit conversions (which is necessary to get BitPat moved out of core) will break rocket-chip. What do you think about deleting the implicit conversions completely in chisel3 (breaks the API, but for good reason) and fixing rocket-chip to use explicit conversions?

sdtwigg commented:
I was willing to punt to a separate PR just to get some motion but your
proposal likely will require the least work overall.

Even if RocketChip uses BitPat extensively, I'd be somewhat surprised if it
was using (and needed to use) the implicit conversions extensively.

On Thu, May 26, 2016 at 2:34 PM, Richard Lin <notifications@github.com>
wrote:

> Still need a way to resolve this. We can keep the explicit conversions,
> but removing the implicit conversions (which is necessary to get BitPat
> moved out of core) will break rocket-chip. What do you think about deleting
> the implicit conversions completely in chisel3 (breaks the API, but for
> good reason) and fixing rocket-chip to use explicit conversions?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/ucb-bar/chisel3/pull/197#issuecomment-222001682>
>

ducky64 commented:
@aswaterman will changing rocket to not use implicit BitPat conversions and eliminating those conversion in chisel3 be acceptable?
Also, comments on splitting chisel._ and chisel.util?

aswaterman commented:
I have no problem adding the implicit conversion to Rocket as a stop-gap.

I also have no problem with splitting out util, as long as it is still possible to compile the same code under chisel2 and chisel3.

ducky64 commented:
Compiling the same code under 2 and 3 would require using the old, deprecated `Chisel` package, which would import `chisel.utils` into `Chisel._`
But once you go to chisel3 and `chisel`, there's no going back (unless the name change or a compatibility layer gets backported to chisel2)

aswaterman commented:
Yeah, that makes sense.

ducky64 commented:
Also, as for adding the implicit conversion to Rocket, I think all the conversions will have to be made explicit, otherwise the implicit conversion from UInt to BitPatComparable (which allows BitPat to be broken out of core by adding in comparison with UInt on the lhs using extension methods / "pimp my library") will fail (it will instead choose to do one of the unsafe conversions, as it is doing right now).

aswaterman commented:
That may or may not be a problem.  Let me look into it.

aswaterman commented:
Are the explicit conversions in Chisel2?

ducky64 commented:
I certainly hope so.
Looks like there is a change in case between `BitPatToUInt` (chisel2) and `bitPatToUInt` (chisel3)
https://github.com/ucb-bar/chisel/blob/master/src/main/scala/BitPat.scala
But nothing that can't be worked around.

If you can remove the uses of the implicit, try building against the chisel3 bitpatsplit branch.

aswaterman commented:
Also, I think we should *not* do the BitPatComparable thing.  While you can argue BitPat doesn't logically belong in core, avoiding the required implicit conversion is a good practical reason to include it in core.

aswaterman commented:
Looks like we'd need several dozen explicit conversions... that's kind of ridiculous.  I'm just going to add the implicit conversion to Rocket.

ducky64 commented:
I think a good argument for breaking it out is that it shows an example of how things like this (where UInt is on the lhs of an operator) can be done in Chisel without modifying Chisel itself. Also, if this does get included in core and a reference put in UInt, the question then becomes what else should go in UInt? This probably isn't the last operator that can take UInt as a lhs operand, and we probably can't anticipate all those uses either.

aswaterman commented:
The slippery slope argument can also be used to argue for not including subtraction, because a+(~b)+1 suffices.  It's just a practical matter.

I agree the PML pattern is a good pattern to extend classes, but requiring its use for supported, baked-in functionality is an anti-pattern.

ducky64 commented:
Subtraction is safe in pretty much all cases (won't assert out) and operates on the same type (so PML is unnecessary there).

It seems that because of the number of these implicit conversions in Rocket, there may be use for some kind of facility where something can act like both a BitPat and UInt simultaneously while being more safe. Probably something that needs more (and separate) discussion.

aswaterman commented:
I think we all agree the implicit conversions that fail in common cases are a bad idea.  I'm only making the case for keeping UInt === BitPat.

ducky64 commented:
Yeah, I know, but this could be implemented using the PML pattern which is mostly transparent to users (except those relying on the current unsafe implicit conversions, which I would argue should be refactored out for code style / safety anyways)

aswaterman commented:
To be perfectly clear, they are not unsafe; they're just caught later than you'd like.

ducky64 commented:
I think it's arguable that any implicit that can assert can be called unsafe? Even if it doesn't silently give you an incorrect circuit, it's obviously doing something very unexpected.

aswaterman commented:
In long-running enterprise applications, perhaps, but I'd argue failure at runtime while generating a hardware circuit is perfectly safe.  Just annoying, because it wasn't caught earlier.

I'm prototyping the PML approach to see how much of an effect it has on us.

ducky64 commented:
I've already prototyped the PML approach in the chisel3 bitpatsplit branch (https://github.com/ucb-bar/chisel3/tree/bitpatsplit), so you should be able to try building against it. No pull request yet since it breaks rocket.

aswaterman commented:
Actually, it's fine for us, because the implicit conversions don't need to stack.   I tried your branch, then deleted the implicit conversions, and added the UInt->BitPat one to Rocket.  It works fine.

I'm still shaky on whether or not it's a good design, but I have to drop my practical opposition.

aswaterman commented:
So, I'm fine with merging this PR, once the merge conflicts are resolved.

ducky64 commented:
Interesting that this works in rocket, since I was unable to get it to work with the implicits remaining in BitPat.scala. Perhaps moving them into rocket changed the implicit resolution order to prefer the BitPat PML conversion instead?

Also, it seems PML is used in Scala base in the form of StringOps.

sdtwigg commented:
Imports are first priority. Above stuff in unimported but 'related' objects
On May 26, 2016 9:30 PM, "Richard Lin" <notifications@github.com> wrote:

> Interesting that this works in rocket, since I was unable to get it to
> work with the implicits remaining in BitPat.scala. Perhaps moving them into
> rocket changed the implicit resolution order to prefer the BitPat PML
> conversion instead?
>
> Also, it seems PML is used in Scala base in the form of StringOps.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucb-bar/chisel3/pull/197#issuecomment-222057752>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AA9_-McG6sb8jp59-Lwhy6Wan--wgOtzks5qFnNEgaJpZM4Ijq9h>
> .
>

ducky64 commented:
Can someone also update the rocket submodule in the rocket-chip repo so the automated builds will work with PML BitPat?

aswaterman commented:
We're fighting with an unrelated issue that prevents us from merging (rocket-chip doesn't allow merges unless CI passes).  I'll let you know when it's in there.

ducky64 commented:
Moved utils into a separate `util` package, rocket appears to still build using the `Chisel` compatibility package.

Will move `chisel._` into `chisel.core` next and have a `chisel` package object that curates what is accessible through `import chisel._`. Also acts as a API reference of sorts.

ducky64 commented:
Moved most of chiselFrontend to `chisel.core`
Moved deprecated functions into `chisel.comaptibility`
Added `chisel` package object as a convenience import for user-facing `chisel.core`. In the future, we may also want to pull in some stuff from `chisel.util`, like the math functions, BitPat, and possibly some common bundle types / constructors (like Decoupled)

ucbjrl commented:
Retest this please

ucbjrl commented:
retest this please.

ucbjrl commented:
retest this please.

ucbjrl commented:
retest this please

ucbjrl commented:
retest this please

ucbjrl commented:
retest this please

ucbjrl commented:
retest this please

ducky64 commented:
Tests finally green!
@sdtwigg @aswaterman anything else that needs to be fixed in the code, or can it be merged?

aswaterman commented:
I'm fine with merging.

ducky64 commented:
Merge issues cleared, tests green, rocket still builds; merging at last!

This takes a pretty strict approach of what goes into chisel (you don't get Decoupled and friends without import chisel.utils), but that's something that can be changed in another PR if desired.
